### DISCUSSION TITLE
Nix noob. How do I add rustaceanvim to my home-manager config and set it up correctly

### DISCUSSION THREAD

**Discussion by multivac61** [mrcjkb/rustaceanvim#59] [Q&A] [answered]
[Here's my config.](https://github.com/multivac61/nixos-config) Would love to learn how to add `rustaceanvim` to it 🤩

**Comment by mrcjkb** [comment, marked as answer]
Hey :wave:

`rustaceanvim` is available in [nixpkgs-unstable](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=rustaceanvim) and will be available in the NixOS 23.11 release (which should be out by the end of this month).

It looks like your flake is using the `master` branch of `nixpks`, so it should be available there, too. I.e. you should be able to just add `rustaceanvim` to your [`plugins` list](https://github.com/multivac61/nixos-config/blob/a68e7b7487cd4705397779ba65245ac9d4b50777/shared/home-manager.nix#L265) and you're good to go.

If you want to use the bleeding-edge latest version of this plugin, I also provide a flake that you can add to your flake's inputs:

```nix
rustaceanvim.url = "github:mrcjkb/rustaceanvim";
```

And then you can add `inputs.rustaceanvim.overlays.default` to your NixOS module:

```nix
nixpkgs.overlays = [
  inputs.rustaceanvim.overlays.default
];
```

As you've been using [`rust-tools.nvim` with `lspconfig`](https://github.com/multivac61/nixos-config/blob/a68e7b7487cd4705397779ba65245ac9d4b50777/shared/config/nvim/lua/plugins/rust.lua#L15), you'll have to remove it from your config (or it may conflict with this plugin).
You can keep `lspconfig`, but don't setup `rust_analyzer` with it.

The `rust-tools.nvim` options are pretty much the same as the `rustaceanvim` ones, only that you set them with a `vim.g.rustaceanvim = {...}` table, instead of calling a `setup` function (which is what `lazy.nvim` does under the hood when you pass it an `opts` table). You can put that anywhere in your Neovim config. I just put it in `init.lua`, but you can also put it in `plugin/rustaceanvim.lua` or in a `lazy.nvim` plugin spec's `init` or `config` function.

Breaking changes from `rust-tools.nvim` to `rustaceanvim` are [covered in the changelog](https://github.com/mrcjkb/rustaceanvim/blob/master/CHANGELOG.md#200---2023-10-21).

Hope this helps :smile: 
Please let me know if you run into any issues.

**Reply by jdsee** [reply]
Hi 👋

Thanks a lot for this elaborate answer and the great work on the plugin!

Unfortunately I'm still stuck. My [Neovim-config](https://github.com/jdsee/nix-config/blob/main/home/jdsee/features/nvim/default.nix) is written in Lua even though I'm using Home-Manager on NixOS. I don't really want to migrate that whole thing to Home-Manager.

It seems like rustaceanvim doesn't work properly on NixOS when not installed via Nix. Most LSP features work just fine on local code but some of them (go to definition, hover, etc.) don't work on the standard library and other external sources.

`nix run "github:mrcjkb/rustaceanvim#nvim-minimal-stable"` works perfectly.

Running `minimal.lua` results in the same problem. The following error is shown in the log:

```
[ERROR project_model::workspace] failed fetching data layout for ManifestPath { file: AbsPathBuf("/home/jdsee/projects/rust/ownership/Cargo.toml") } workspace e=RUSTC_BOOTSTRAP="1" "rustc" "-Z" "unstable-options" "--print" "target-spec-json" failed
```

I also tried to add the overlay from the flake, but got the same error there.

I guess this issue could be related to the dynamic linker but have absolutely no idea how to further debug it. Is it possible to integrate rustaceanvim via Nix without needing to migrate my whole config to Home-Manager?


Any ideas or help is very much appreciated 🙏

**Reply by mrcjkb** [reply]
@jdsee this looks like a toolchain issue (I think the error message you're seeing comes from rust-analyzer).

In your config, you add rust-analyzer to your `home.packages`, but no other rust tools. You also need cargo and rustc.

If you run `:checkhealth rustaceanvim`, it should tell you what is missing 😄

**Reply by jdsee** [reply]
Oh wow, now I feel silly 😄

I have cargo installed in another place and so far used nix-shells to pull in rust toolchain. Unfortunately I didn’t have a nix-shell in the project that I used to try out rustacaenvim.

Thanks a bunch for the quick reply and your help 🙏 I’m completely new to rust and this plugin as well as your support are a huge motivation!

**Comment by damccull** [comment]
Solved my toolchain issue too. :D And I am using kickstart-nix.nvim.
